docs/guides/validation-user-guide.md

```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
+ - `spine.time.LocalDate`;
+ - `spine.time.LocalDateTime`;
+ - `spine.time.OffsetDateTime`;
+ - `spine.time.ZonedDateTime`;
+ - any user-defined type which implements the Temporal interface (`io.spine.time.Temporal` for
+   Java).
+
+Using the option `(when)`, you may declare that the timestamp should lie in past or in future.
+
+```proto
+import "spine/time_options.proto";
+import "spine/time/time.proto";
+
+// A command to place an order.
+message PlaceOrder {
+    
+    // ...
+
+    spine.time.ZonedDateTime when_placed = 12 [(when).in = PAST];
+    spine.time.ZonedDateTime when_expires = 13 [(when).in = FUTURE];
+}
+```
+
+Note that the value is checked in relation to the current server time. In most cases, this should
+not be an issue. However, be aware that using `FUTURE` in Events and entity states may cause
+validation errors upon playing historical Events in future. However, that is not the case with
+Commands.  
+
+### Distinct collections
+
+Often, a `repeated` field logically represents a set rather than a list. Protobuf does not have
+a native support for sets. Moreover, it is often an invalid operation to add a duplicate element to
+a set. For such cases, Spine provides the `(distinct)` option, which constrains a `repeated` or
+a `map` field to only contain non-duplicating elements (values in case of `map`s).
+
+Example:
+
+```proto
+import "spine/options.proto";
+import "spine/net/email_address.proto";
+
+// The state of the User Aggregate.
+message User {
+    // ...
+
+    repeated spine.net.EmailAddress recovery_emails = 42 [(distict) = true];
+}
+```
+
+### Non-mutable state fields
+
+For entity states, Spine defines a special validation constraint. It is not typically checked in any
+other situation, but only when updating a state of an existing Spine entity.
+
+Many fields of an entity are immutable. They may be set one in the life of the entity and then
+should never be changed. In order to enforce this, we provide the `(set_once)` option. The option
+allows changing a value of a field only if the current value is the default value. Changing a field
+from a non-default value to anything else will cause a violation.
+
+Example:
+
+```proto
+import "spine/options.proto";
+import "google/protobuf/timestamp.proto";
+
+// The state of the Order Aggregate.
+message Order {
+    option (entity).kind = AGGREGATE;
+
+    // ...
+    
+    google.protobuf.Timestamp when_deleted = 314 [(set_once) = true];
+}
+```
+
+Once the `Order.when_deleted` field is filled, it can never change.
+
+## External constraints
+
+Sometimes, you need to impose extra validation rules on types you do not control. Consider
+the example of an image URL which should always have the `ftp` protocol:
+
+```proto
+import "spine/net/url.proto";
+
+// The state of the User Aggregate.
+message User {
+    // ...
+
+    spine.net.Url profile_picture = 42;
+}
+```
+
+How do we add validation to the fields inside `Url` so that only the `User.profile_picture` is
```

Comment:
Let's also show how `Url` looks. We don't expect readers to take their time and go look into the definition themselves, do we?

It makes even more difference, given you speak of `Url` structure further, implying it's a well-known thing.